src/libhook/libhook.hpp

```diff
@@ -0,0 +1,150 @@
+ * If you have received a written license agreement or contract for        *
+ * Covered Software stating terms other than these, you may choose to use  *
+ * and redistribute Covered Software under those terms instead of these.   *
+ *                                                                         *
+ * Source is provided to this software because we believe users have a     *
+ * right to know exactly what a program is going to do before they run it. *
+ * This also allows you to audit the software for security holes.          *
+ *                                                                         *
+ * Source code also allows you to port DRAKVUF to new platforms, fix bugs, *
+ * and add new features.  You are highly encouraged to submit your changes *
+ * on https://github.com/tklengyel/drakvuf, or by other methods.           *
+ * By sending these changes, it is understood (unless you specify          *
+ * otherwise) that you are offering unlimited, non-exclusive right to      *
+ * reuse, modify, and relicense the code.  DRAKVUF will always be          *
+ * available Open Source, but this is important because the inability to   *
+ * relicense code has caused devastating problems for other Free Software  *
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ ***************************************************************************/
+#pragma once
+
+#include "call_result.hpp"
+#include "hooks/base.hpp"
+#include "hooks/manual.hpp"
+#include "hooks/return.hpp"
+#include "hooks/syscall.hpp"
+
+/**
+ * A brief information about caveats of libhook.
+ *
+ * There are few things libhook provides:
+ *  1) RAII containers around libdrakvuf traps
+ *  2) Trap API unification
+ *  3) ability to use non-static class-member-functions as callbacks
+ *
+ * 1) I spent over 2 days debugging this issue, so I'm going to note it down here (for future reference).
+ *
+ * If we delete a hook, we call dctor of the hook object, where we call drakvuf_remove_trap.
+ * The problem is that drakvuf doesn't immediately remove those traps, but waits for entire drakvuf loop
+ * to pass. This means that drakvuf might call hook, which has been already free'd.
```

Comment:
That's what `drakvuf_trap_free_t` is for in `drakvuf_remove_trap`. So you only `free` things when you know its safe.